### ISSUE TITLE
Within-epoch surrogate generation

### ISSUE THREAD

**Issue by seqasim** [mne-tools/mne-connectivity#251]
#### Describe the problem

It is difficult to interpret many connectivity estimates without comparison to "null" surrogate data. As stated in [Aru et al., 2015](https://www.sciencedirect.com/science/article/pii/S0959438814001640#bib0015) : "A suitable surrogate construction should only destroy the specific cyclo-stationarities related to the hypothesized... effect, while keeping all the unspecific non-stationarities and non-linearities of the original data."

#### Describe your solution
It would make sense to perform trial/epoch level scrambling so that non-stationarities and phase distributions within a trial/epoch are conserved. [Tensorpac](https://github.com/EtienneCmb/tensorpac/blob/ac9058fd375d423757943810f613d63785fab85f/tensorpac/methods/meth_surrogates.py#L59) achieves this with a simple function, which I have edited slightly:

```
def swap_time_blocks(data, random_state=None):

    """Compute surrogates by swapping time blocks.
    This function cuts the timeseries at a random time point. Then, both time
    blocks are swapped.
    Parameters
    ----------
    data : array_like
        Array of shape (n_chan, ..., n_times).
    random_state : int | None
        Fix the random state of the machine for reproducible results.
    Returns
    -------
    surr : array_like
        Swapped timeseries to use to compute the distribution of
        permutations
    References
    ----------
    Source: https://www.sciencedirect.com/science/article/pii/S0959438814001640
    """
    
    if random_state is None:
        random_state = int(np.random.randint(0, 10000, size=1))
    rnd = np.random.RandomState(random_state)
    
    # get the minimum / maximum shift
    min_shift, max_shift = 1, None
    if not isinstance(max_shift, (int, float)):
        max_shift = data.shape[-1]
    # random cutting point along time axis
    cut_at = rnd.randint(min_shift, max_shift, (1,))
    # split amplitude across time into two parts
    surr = np.array_split(data, cut_at, axis=-1)
    # revered elements
    surr.reverse()
    
    return np.concatenate(surr, axis=-1)
```

Example usage with mne data: 

```
    data = np.swapaxes(mne_data.get_data(copy=False), 0, 1) # swap so now it's chan, events, times 
    surr_dat = np.zeros_like(data) # allocate space for the surrogate channels 
    for ix, ch_dat in enumerate(data): # apply the same swap to every event in a channel, but differ between channels 
        surr_ch = swap_time_blocks(ch_dat, random_state=None)
        surr_dat[ix, :, :] = surr_ch
    surr_dat = np.swapaxes(surr_dat, 0, 1) # swap back so it's events, chan, times 
    # make a new EpochArray from it
    surr_mne = mne.EpochsArray(surr_dat, 
                mne_data.info, 
                tmin=mne_data.tmin, 
                events = mne_data.events, 
                event_id = mne_data.event_id)
```

**Comment by tsbinns** [comment]
Hi @seqasim,

I wonder how much this need is addressed with this PR I wrote over the summer: https://github.com/mne-tools/mne-connectivity/pull/223

It has a similar purpose to generate some surrogate estimates of null connectivity based on the procedure (briefly) described in these publications: https://www.sciencedirect.com/science/article/pii/S1053811923003695; https://www.biorxiv.org/content/10.1101/2023.10.26.564193v1.abstract

This example demonstrates things: https://output.circle-artifacts.com/output/job/bafe351e-3009-46ea-92ce-863de6768196/artifacts/0/dev/auto_examples/surrogate_connectivity.html#sphx-glr-auto-examples-surrogate-connectivity-py

Cheers!

**Comment by seqasim** [comment]
Yes, I saw this nice addition! But I think the example you posted highlights the need for intra-trial shuffling:

"Critically, evoked data contains a temporal structure that is consistent across epochs, and thus shuffling epochs across channels will fail to adequately disrupt the covariance structure." 

I believe that cutting the data within the trial, as I propose, would deal with this issue?

**Comment by tsbinns** [comment]
Cutting just once within the trial is not something we had looked into before.

I adapted the code you wrote into a function that performs this shuffling multiple times and incorporated it into the example I linked to:
![image](https://github.com/user-attachments/assets/99b1e204-4908-4fa4-a7fa-2d50ff5b5d58)
Orange is the connectivity for the evoked response, red is the connectivity for the surrogate with the proposed within-epoch shuffling, and blue is the connectivity from the pre-stimulus surrogate with the existing procedure in that PR.

It seems there is still a lot of residual connectivity information preserved using the within-epoch shuffling. Perhaps also cutting the data independently for each epoch would reduce this. Making additional cuts within each epoch could be tricky, as then you move towards creating white noise.

I don't think it's due to an error when I adapted the code, but here it is in case I missed something:
```
def make_surrogate_data_alt(
    data, n_shuffles=1000, rng_seed=None, return_generator=True
):
    """Create surrogate data for a null hypothesis of connectivity."""
    surrogate = _shuffle_within_epochs(data, n_shuffles, rng_seed)
    if not return_generator:
        surrogate = [shuffle for shuffle in surrogate]

    return surrogate


def _shuffle_within_epochs(data, n_shuffles, rng_seed):
    """Shuffle within epochs in data."""
    from mne import EpochsArray

    data_arr = data.get_data(copy=True)
    rng = np.random.default_rng(rng_seed)
    for _ in range(n_shuffles):
        surr_arr = np.zeros_like(data_arr)
        for ch_idx in range(data_arr.shape[1]):
            surr_arr[:, ch_idx] = _swap_time_blocks(data_arr[:, ch_idx], rng)
        yield EpochsArray(surr_arr, info=data.info)


def _swap_time_blocks(data, rng):
    """Swap time blocks in data."""
    min_shift = 1
    max_shift = data.shape[-1]
    cut_at = rng.integers(min_shift, max_shift, 1)
    surr = np.array_split(data, cut_at, axis=-1)
    surr.reverse()
    return np.concatenate(surr, axis=-1)
```

**Comment by seqasim** [comment]
Wow, thanks for adapting this so quickly! Lovely to see a head-to-head comparison.

Based on this real-data example, though, it's not clear to me whether the pre-stimulus surrogate (blue) will give false positive assessments or my post-stimulus surrogate (red) will give false negative assessments - maybe the true task-evoked effect in this MEG data is truly in the 6-12 Hz range? 

Would it perhaps be better to simulate data with varying levels of frequency-specific power and true connectivity and test which surrogate method recovers ground truth most effectively? 

Or at least it might make sense to provide users with all three options: pre-stimulus epoch shuffle (most permissive), post-stimulus epoch shuffle (medium permissive), post-stimulus cut and swap (least permissive) ?

**Comment by tsbinns** [comment]
> Would it perhaps be better to simulate data with varying levels of frequency-specific power and true connectivity and test which surrogate method recovers ground truth most effectively?

Definitely. For the between channel epoch shuffling, this was tested using data from the [`make_surrogate_connectivity()`](https://mne.tools/mne-connectivity/stable/generated/mne_connectivity.make_signals_in_freq_bands.html#mne_connectivity.make_signals_in_freq_bands) function, which uses a pretty standard approach of generating spectral connectivity data. However, the data from this should be considered more as resting state data. This approach would need to be adapted to generate connectivity for evoked data.

Still, even for resting state data the within epoch shuffling would need to be able to identify this baseline.

**Comment by seqasim** [comment]
Ah I see. I think you're right about cutting the data independently for every epoch for evoked data. I modified your code to do so. Could you possibly test this with the example you provided above? 


```
import numpy as np
from mne import EpochsArray

def make_surrogate_data_alt(data, n_shuffles=1000, rng_seed=None, return_generator=True):
    """Create surrogate data for a null hypothesis of connectivity."""
    surrogate = _shuffle_within_epochs(data, n_shuffles, rng_seed)
    if not return_generator:
        surrogate = [shuffle for shuffle in surrogate]
    return surrogate

def _shuffle_within_epochs(data, n_shuffles, rng_seed):
    """Shuffle within epochs in data."""
    data_arr = data.get_data(copy=True)
    rng = np.random.default_rng(rng_seed)
    for _ in range(n_shuffles):
        surr_arr = np.zeros_like(data_arr)
        cutpoints = rng.integers(1, data_arr.shape[-1], (data_arr.shape[0], data_arr.shape[1]))
        for ev_idx in range(data_arr.shape[0]):
            for ch_idx in range(data_arr.shape[1]):
                surr_arr[ev_idx, ch_idx] = _swap_time_blocks(data_arr[ev_idx, ch_idx], cutpoints[ev_idx, ch_idx])
        yield EpochsArray(surr_arr, info=data.info)

def _swap_time_blocks(data, cut_at):
    """Swap time blocks in data at a given cutpoint."""
    surr = np.array_split(data, [cut_at], axis=-1)
    surr.reverse()
    return np.concatenate(surr, axis=-1)
```

Note that `cutpoints` should on contain a unique cut for each event, and and each channel. This should preserve within-trial dynamics, but destroy across-trial residual connectivity. Thoughts?

**Comment by tsbinns** [comment]
This is what I get with the code you modified (again, blue is existing method; red is suggested method):
![image](https://github.com/user-attachments/assets/1df265b9-a970-4f18-b76a-dafcd70ade1f)
The additional independent shuffling for each epoch definitely helps.

While this new approach is somewhat simpler for working with evoked data, it is more computationally demanding since you need to compute the Fourier coefficients for each shuffle of the timeseries, whereas the approach in the existing PR allows you to compute the Fourier coefficients once and then shuffle these.

I would be interested to hear the thoughts of @larsoner, @wmvanvliet, and @drammock, and if there's a way both approaches could be made available.

**Comment by larsoner** [comment]
I don't have time to dig into the details at the moment, but given https://www.sciencedirect.com/science/article/pii/S1053811923003695 is published by knowledgeable folks in a good journal, assuming you're suggesting to make methods from that available then it sounds reasonable to me!

**Comment by tsbinns** [comment]
Was looking over some of the open PRs with @drammock, and we think this would be a nice addition. CC @seqasim @larsoner

> given https://www.sciencedirect.com/science/article/pii/S1053811923003695 is published by knowledgeable folks in a good journal, assuming you're suggesting to make methods from that available then it sounds reasonable to me!

That paper shows the methods for surrogate generation we already have implemented. This paper cited by tensorpac proposes the alternative method: https://www.sciencedirect.com/science/article/pii/S0959438814001640

The existing surrogate data generation works great for resting-state data, and while the approach can be used for assessing significant connectivity in evoked data:
1) this is practically more cumbersome; and
2) sometimes this might not even be possible (e.g., if you're working with datasets where only derivatives are available, such that there isn't any resting state data or a sufficient amount of pre-trial data).

There is the question of what the API looks like. An important consideration is what forms of data each approach can create surrogates from.

The existing method is compatible with (non-)time-resolved data (i.e., `EpochsSpectrum` and `EpochsTFR`). Because this alternative method shuffles timepoints within epochs, **it can't work** with non-time-resolved data (i.e., `EpochsSpectrum`). So, really the new method would need to support `Epochs` data, which you shuffle and then compute spectral coefficients from for each shuffle (assuming you are estimating connectivity from `EpochsSpectrum`, otherwise `EpochsTFR` could be used directly).

Support for `Epochs` wasn't added to the existing approach when this was first implemented, as I only considered this from the perspective of spectral connectivity. It is computationally cheaper to compute spectral coefficients once, shuffle these, and then compute connectivity, rather then shuffle `Epochs` data and re-compute spectral coefficients for each shuffle. However, I now realise this ignores cases where surrogates for non-spectral connectivity is desired, so I think the existing approach should be extended to allow `Epochs` (wouldn't involve methods changes beyond allowing the func to accept `Epochs` data).

The implementation of the alternative approach could come as an inclusion in `make_surrogate_data` as a `method`-like parameter. The existing implementation could remain as the default behaviour, accepting `Epochs`, `EpochsSpectrum`, and `EpochsTFR` data. The alternative method could also be selected, accepting `Epochs` and `EpochsTFR` data.

There is a risk of the wrong approach being used. Maybe we could warn if it looks like the data has been epoched around some event, e.g., it has negative timestamps, or was baselined from data with negative timestamps (that were since cropped)? Or, maybe after a depr cycle there is no default method, and users must specify a `data_form`-like parameter for `"resting"` or `"evoked"`.

Otherwise, there could be two separate functions, e.g., `make_surrogate_data` → `make_surrogate_resting_data` for existing functionality, `make_surrogate_evoked_data` for alternative method. I expect this has less risk of the wrong approach being used.

**Comment by larsoner** [comment]
I lean toward the two different function names, since it seems like (1) it actually uses a different shuffling mechanism under the hood, and (2) it reduces the risk of incorrect usage
